File: src/testing/adapters.ts
import type { FetchAdapter, FetchResponse, SearchAdapter, SearchHit } from '../types.js';
import { compareCLocaleText, tokenizeForSearch } from '../utils.js';

export interface DemoSearchDocument {
  title: string;
  body: string;
  source: string;
}

interface ScoredSearchHit {
  score: number;
  hit: SearchHit;
}

export class DemoSearch implements SearchAdapter {
  private readonly docs: DemoSearchDocument[];

  constructor(docs: DemoSearchDocument[]) {
    this.docs = docs.map(cloneDemoSearchDocument);
  }

  async search(query: string, limit = 10): Promise<SearchHit[]> {
    const queryTokens = tokenizeForSearch(query);
    const scored: ScoredSearchHit[] = [];

    for (const doc of this.docs) {
      const overlap = countTokenOverlap(queryTokens, `${doc.title}\n${doc.body}`);
      if (overlap === 0) {
        continue;
      }

      scored.push({
        score: overlap,
        hit: {
          title: doc.title,
          snippet: truncateSnippet(doc.body),
          source: doc.source,
        },
      });
    }

    scored.sort(compareScoredSearchHits);
    return scored.slice(0, limit).map(function (entry) {
      return entry.hit;
    });
  }
}

export class DemoFetch implements FetchAdapter {
  private readonly resources: Record<string, unknown>;

  constructor(resources: Record<string, unknown>) {
    this.resources = cloneResourceMap(resources);
  }

  async fetch(resource: string): Promise<FetchResponse> {
    if (!Object.hasOwn(this.resources, resource)) {
      throw new Error(`resource not found: ${resource}`);
    }

    const payload = cloneDemoPayload(this.resources[resource]);
    if (payload instanceof Uint8Array) {
      return {
        contentType: 'application/octet-stream',
        payload,
      };
    }
    if (typeof payload === 'string') {
      return {
        contentType: 'text/plain',
        payload,
      };
    }
    return {
      contentType: 'application/json',
      payload,
    };
  }
}

function cloneDemoSearchDocument(doc: DemoSearchDocument): DemoSearchDocument {
  return {
    title: doc.title,
    body: doc.body,
    source: doc.source,
  };
}

function countTokenOverlap(queryTokens: Set<string>, text: string): number {
  const tokens = tokenizeForSearch(text);
  let overlap = 0;

  for (const token of queryTokens) {
    if (tokens.has(token)) {
      overlap += 1;
    }
  }

  return overlap;
}

function truncateSnippet(body: string): string {
  const snippet = body.replace(/\s+/g, ' ').trim();
  if (snippet.length <= 160) {
    return snippet;
  }
  return `${snippet.slice(0, 157)}...`;
}

function compareScoredSearchHits(left: ScoredSearchHit, right: ScoredSearchHit): number {
  if (right.score !== left.score) {
    return right.score - left.score;
  }
  return compareCLocaleText(left.hit.title, right.hit.title);
}

function cloneResourceMap(resources: Record<string, unknown>): Record<string, unknown> {
  const clone: Record<string, unknown> = {};

  for (const [resource, payload] of Object.entries(resources)) {
    clone[resource] = cloneDemoPayload(payload);
  }

  return clone;
}

function cloneDemoPayload(payload: unknown): unknown {
  if (payload instanceof Uint8Array) {
    return new Uint8Array(payload);
  }

  if (
    payload === null ||
    payload === undefined ||
    typeof payload === 'string' ||
    typeof payload === 'number' ||
    typeof payload === 'boolean'
  ) {
    return payload;
  }

  return structuredClone(payload);
}

File: python/src/onetool/testing/adapters.py
from __future__ import annotations

from copy import deepcopy
from dataclasses import dataclass
from typing import Any

from ..c_locale import to_c_locale_sort_bytes
from ..types import FetchResponse, SearchHit
from ..utils import tokenize_for_search


@dataclass(frozen=True, slots=True)
class DemoSearchDocument:
    title: str
    body: str
    source: str


@dataclass(frozen=True, slots=True)
class _ScoredSearchHit:
    score: int
    hit: SearchHit


class DemoSearch:
    def __init__(self, docs: list[DemoSearchDocument]) -> None:
        self._docs = [DemoSearchDocument(doc.title, doc.body, doc.source) for doc in docs]

    async def search(self, query: str, limit: int = 10) -> list[SearchHit]:
        query_tokens = tokenize_for_search(query)
        scored: list[_ScoredSearchHit] = []

        for doc in self._docs:
            overlap = _count_token_overlap(query_tokens, f"{doc.title}\n{doc.body}")
            if overlap == 0:
                continue

            scored.append(
                _ScoredSearchHit(
                    score=overlap,
                    hit=SearchHit(
                        title=doc.title,
                        snippet=_truncate_snippet(doc.body),
                        source=doc.source,
                    ),
                )
            )

        scored.sort(key=lambda entry: (-entry.score, to_c_locale_sort_bytes(entry.hit.title)))
        return [entry.hit for entry in scored[:limit]]


class DemoFetch:
    def __init__(self, resources: dict[str, Any]) -> None:
        self._resources = {resource: _clone_demo_payload(payload) for resource, payload in resources.items()}

    async def fetch(self, resource: str) -> FetchResponse:
        if resource not in self._resources:
            raise ValueError(f"resource not found: {resource}")

        payload = _clone_demo_payload(self._resources[resource])
        if isinstance(payload, bytes):
            return FetchResponse(content_type="application/octet-stream", payload=payload)
        if isinstance(payload, str):
            return FetchResponse(content_type="text/plain", payload=payload)
        return FetchResponse(content_type="application/json", payload=payload)


def _count_token_overlap(query_tokens: set[str], text: str) -> int:
    tokens = tokenize_for_search(text)
    overlap = 0
    for token in query_tokens:
        if token in tokens:
            overlap += 1
    return overlap


def _truncate_snippet(body: str) -> str:
    snippet = " ".join(body.split())
    if len(snippet) <= 160:
        return snippet
    return f"{snippet[:157]}..."


def _clone_demo_payload(payload: Any) -> Any:
    if isinstance(payload, bytes):
        return bytes(payload)
    if isinstance(payload, bytearray):
        return bytes(payload)
    if payload is None or isinstance(payload, (str, int, float, bool)):
        return payload
    return deepcopy(payload)
